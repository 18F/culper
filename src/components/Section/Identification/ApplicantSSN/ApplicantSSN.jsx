import React from 'react'
import { i18n } from '../../../../config'
import schema from '../../../../schema'
import validate from '../../../../validators'
import { validSSN } from '../../../../validators/helpers'
import SubsectionElement from '../../SubsectionElement'
import { Field, SSN, Show } from '../../../Form'

export default class ApplicantSSN extends SubsectionElement {
  constructor (props) {
    super(props)

    this.state = {
      uid: `${this.props.name}-${super.guid()}`,
      verification: {},
      error: false
    }

    this.updateSSN = this.updateSSN.bind(this)
    this.updateVerification = this.updateVerification.bind(this)
    this.verificationError = this.verificationError.bind(this)
  }

  update (queue) {
    this.props.onUpdate({
      ssn: this.props.ssn,
      verified: this.props.verified,
      ...queue
    })
  }

  updateSSN (values) {
    this.update({
      ssn: values,
      verified: false,
      error: false
    })
  }

  updateVerification (values) {
    const verified = this.props.ssn.first === values.first &&
          this.props.ssn.middle === values.middle &&
          this.props.ssn.last === values.last &&
          validSSN(values)

    this.setState({
      verification: verified ? {} : values,
      error: verified ? false : this.state.error
    }, () => {
      // Update the properties so they can be reflected
      // within this component.
      this.update({ verified: verified })

      // If everything checks out there are no errors and
      // we are going to forcefully flush them.
      if (verified) {
        this.handleError(values, [])
        this.verificationError(values, [])
      }
    })
  }

  verificationError (value, arr) {
    // If the verification SSN is valid then we run our validation
    // function(s). However, if it is not then we set the `valid` property
    // to a neutral state. This allows the error notifications to be properly
    // toggled within the various subscribing components.
    const verification = this.state.verification
    const local = this.props.onError(value, this.constructor.errors.map(err => {
      return {
        code: err.code,
        valid: validSSN(verification) ? err.func(verification, this.props) : null,
        uid: this.state.uid
      }
    })) || []

    // Set the error state value so we can apply a CSS class to
    // the entire SSN.
    this.setState({ error: local.some(x => x.valid === false) })
    return this.handleError(value, arr.concat(local))
  }

  render () {
    const klass = `section-content applicant-ssn ${this.props.className || ''}`.trim()
    const klassVerify = `applicant-ssn-verification ${this.state.error ? 'usa-input-error' : ''}`.trim()
    const verify = validSSN(this.props.ssn) && !this.props.verified && !this.props.ssn.notApplicable

    return (
      <div className={klass} {...super.dataAttributes(this.props)}>
        <Field title={i18n.t('identification.ssn.title')}
               titleSize="h2"
               help="identification.ssn.help"
               scrollIntoView={this.props.scrollIntoView}>
          <SSN name="ssn"
               {...this.props.ssn}
               className="applicant-ssn-initial"
               onUpdate={this.updateSSN}
               onError={this.handleError}
               required={!this.props.ssn.notApplicable && this.props.required}
               />
        </Field>

        <Show when={verify}>
          <Field title={i18n.t('identification.ssn.heading.verify')}
                 scrollIntoView={this.props.scrollIntoView}
                 titleSize="h4">
            <SSN name="verification"
                 {...this.state.verification}
                 hideNotApplicable={true}
                 className={klassVerify}
                 onUpdate={this.updateVerification}
                 onError={this.verificationError}
                 required={verify && this.props.required}
                 />
          </Field>
        </Show>
        <Show when={this.props.verified}>
          <Field title={i18n.t('identification.ssn.heading.verified')}
                 optional={true}
                 scrollIntoView={this.props.scrollIntoView}
                 titleSize="h4">
          </Field>
        </Show>
      </div>
    )
  }
}

ApplicantSSN.defaultProps = {
  name: 'applicant-ssn',
  ssn: {},
  verified: false,
  onUpdate: (queue) => {},
  onError: (value, arr) => { return arr },
  section: 'identification',
  subsection: 'ssn',
  dispatch: () => {},
  required: false,
  validator: (data) => {
    return validate(schema('identification.ssn', data))
  }
}

ApplicantSSN.errors = [
  {
    code: 'ssn.mismatch',
    func: (value, props) => {
      if (!value) {
        return null
      }
      return validate(schema('ssn', value)) &&
        props.ssn.first === value.first &&
        props.ssn.middle === value.middle &&
        props.ssn.last === value.last
    }
  }
]
