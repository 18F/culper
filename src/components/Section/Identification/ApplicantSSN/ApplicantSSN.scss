.ssn {
    .first {
        width: 9rem !important;
        display: inline-block;

        input {}
    }

    .middle {
        width: 8rem !important;
        display: inline-block;
        margin-left: 2rem;
        margin-right: 2rem;

        input {}
    }

    .last {
        width: 10rem !important;
        display: inline-block;

        input {}
    }

    .coupled-flags {
        padding-left: 18rem;
    }

    .toggle {
        margin-top: 2.6rem;
    }
}
