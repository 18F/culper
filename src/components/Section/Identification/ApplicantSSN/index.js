import ApplicantSSN from './ApplicantSSN'
export default ApplicantSSN
