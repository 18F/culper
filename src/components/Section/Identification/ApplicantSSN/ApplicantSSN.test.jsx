import React from 'react'
import { shallow, mount } from 'enzyme'
import { ApplicantSSN } from './ApplicantSSN'

// give a fake GUID so the field IDs don't differ between snapshots
jest.mock('components/Form/ValidationElement/index')

describe('The ApplicantSSN component', () => {
  it('renders without errors', () => {
    const component = shallow(<ApplicantSSN />)

    expect(component.exists()).toBe(true)
    expect(component).toMatchSnapshot()
  })

  it('implements an onUpdate handler', () => {
    let updates = 0
    const testProps = {
      onUpdate: () => {
        updates += 1
      },
    }

    const component = mount(<ApplicantSSN {...testProps} />)
    component.find('.applicant-ssn-initial .first input').simulate('change')
    expect(updates).toBe(1)
  })

  describe('initial state', () => {
    it('asks for only one SSN', () => {
      const component = mount(<ApplicantSSN />)
      expect(component.find('.applicant-ssn-initial').length).toBe(1)
      expect(component.find('.applicant-ssn-verification').length).toBe(0)
    })
  })

  describe('with a value, unverified', () => {
    it('asks for verification', () => {
      const props = {
        ssn: {
          first: '123',
          middle: '12',
          last: '1234',
        },
        verified: false,
      }

      const component = mount(<ApplicantSSN {...props} />)
      expect(component.find('.applicant-ssn-initial').length).toBe(1)
      expect(component.find('.applicant-ssn-verification').length).toBe(1)
    })
  })

  describe('with a value, verified', () => {
    it('does not ask to be verified again', () => {
      const props = {
        ssn: {
          first: '123',
          middle: '12',
          last: '1234',
        },
        verified: true,
      }

      const component = mount(<ApplicantSSN {...props} />)
      expect(component.find('.applicant-ssn-initial').length).toBe(1)
      expect(component.find('.applicant-ssn-verification').length).toBe(0)
    })

    it('sets verified to false if the initial SSN is edited', () => {
      let dirty = false
      const props = {
        ssn: {
          first: '123',
          middle: '12',
          last: '1234',
        },
        verified: true,
        onUpdate: (queue) => {
          dirty = !queue.verified
        },
      }

      const component = mount(<ApplicantSSN {...props} />)
      expect(component.find('.applicant-ssn-initial').length).toBe(1)
      expect(component.find('.applicant-ssn-verification').length).toBe(0)
      component.find('.applicant-ssn-initial .first input').simulate('change')
      expect(dirty).toBe(true)
    })
  })
})
