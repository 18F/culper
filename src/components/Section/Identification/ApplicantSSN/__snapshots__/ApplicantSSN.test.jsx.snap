// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The applicant SSN component renders without errors 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="section-content applicant-ssn"
    data-section="identification"
    data-subsection="ssn"
  >
    <h1
      className="section-header"
    >
      Social security number
    </h1>
    <Field
      adjustFor=""
      className=""
      comments={false}
      commentsActive={false}
      commentsAdd="comments.add"
      commentsName="Comments"
      commentsRemove="comments.remove"
      commentsRequired={false}
      commentsValue={Object {}}
      errorPrefix=""
      errors={Array []}
      filterErrors={[Function]}
      help="identification.ssn.help"
      helpActive={false}
      optional={false}
      optionalText=""
      scrollIntoView={true}
      shrink={false}
      title="Provide your U.S. Social Security Number"
      titleSize="h4"
      validate={true}
    >
      <SSN
        className="applicant-ssn-initial"
        error={false}
        first=""
        focus={false}
        hideNotApplicable={false}
        last=""
        middle=""
        name="ssn"
        notApplicable={false}
        onError={[Function]}
        onUpdate={[Function]}
        required={false}
        tab={[Function]}
        valid={false}
        value=""
      />
    </Field>
    <Show
      when={false}
    >
      <Field
        adjustFor=""
        className=""
        comments={false}
        commentsActive={false}
        commentsAdd="comments.add"
        commentsName="Comments"
        commentsRemove="comments.remove"
        commentsRequired={false}
        commentsValue={Object {}}
        errorPrefix=""
        errors={Array []}
        filterErrors={[Function]}
        help=""
        helpActive={false}
        optional={false}
        optionalText=""
        scrollIntoView={true}
        shrink={false}
        title="Please confirm your social security number"
        titleSize="h4"
        validate={true}
      >
        <SSN
          className="applicant-ssn-verification"
          error={false}
          first=""
          focus={false}
          hideNotApplicable={true}
          last=""
          middle=""
          name="verification"
          notApplicable={false}
          onError={[Function]}
          onUpdate={[Function]}
          required={false}
          tab={[Function]}
          valid={false}
          value=""
        />
      </Field>
    </Show>
    <Show
      when={false}
    >
      <Field
        adjustFor=""
        className=""
        comments={false}
        commentsActive={false}
        commentsAdd="comments.add"
        commentsName="Comments"
        commentsRemove="comments.remove"
        commentsRequired={false}
        commentsValue={Object {}}
        errorPrefix=""
        errors={Array []}
        filterErrors={[Function]}
        help=""
        helpActive={false}
        optional={true}
        optionalText=""
        scrollIntoView={true}
        shrink={false}
        title="Thank you for confirming your social security number!"
        titleSize="h4"
        validate={true}
      />
    </Show>
  </div>,
  "nodes": Array [
    <div
      className="section-content applicant-ssn"
      data-section="identification"
      data-subsection="ssn"
    >
      <h1
        className="section-header"
      >
        Social security number
      </h1>
      <Field
        adjustFor=""
        className=""
        comments={false}
        commentsActive={false}
        commentsAdd="comments.add"
        commentsName="Comments"
        commentsRemove="comments.remove"
        commentsRequired={false}
        commentsValue={Object {}}
        errorPrefix=""
        errors={Array []}
        filterErrors={[Function]}
        help="identification.ssn.help"
        helpActive={false}
        optional={false}
        optionalText=""
        scrollIntoView={true}
        shrink={false}
        title="Provide your U.S. Social Security Number"
        titleSize="h4"
        validate={true}
      >
        <SSN
          className="applicant-ssn-initial"
          error={false}
          first=""
          focus={false}
          hideNotApplicable={false}
          last=""
          middle=""
          name="ssn"
          notApplicable={false}
          onError={[Function]}
          onUpdate={[Function]}
          required={false}
          tab={[Function]}
          valid={false}
          value=""
        />
      </Field>
      <Show
        when={false}
      >
        <Field
          adjustFor=""
          className=""
          comments={false}
          commentsActive={false}
          commentsAdd="comments.add"
          commentsName="Comments"
          commentsRemove="comments.remove"
          commentsRequired={false}
          commentsValue={Object {}}
          errorPrefix=""
          errors={Array []}
          filterErrors={[Function]}
          help=""
          helpActive={false}
          optional={false}
          optionalText=""
          scrollIntoView={true}
          shrink={false}
          title="Please confirm your social security number"
          titleSize="h4"
          validate={true}
        >
          <SSN
            className="applicant-ssn-verification"
            error={false}
            first=""
            focus={false}
            hideNotApplicable={true}
            last=""
            middle=""
            name="verification"
            notApplicable={false}
            onError={[Function]}
            onUpdate={[Function]}
            required={false}
            tab={[Function]}
            valid={false}
            value=""
          />
        </Field>
      </Show>
      <Show
        when={false}
      >
        <Field
          adjustFor=""
          className=""
          comments={false}
          commentsActive={false}
          commentsAdd="comments.add"
          commentsName="Comments"
          commentsRemove="comments.remove"
          commentsRequired={false}
          commentsValue={Object {}}
          errorPrefix=""
          errors={Array []}
          filterErrors={[Function]}
          help=""
          helpActive={false}
          optional={true}
          optionalText=""
          scrollIntoView={true}
          shrink={false}
          title="Thank you for confirming your social security number!"
          titleSize="h4"
          validate={true}
        />
      </Show>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ApplicantSSN
        dispatch={[Function]}
        onError={[Function]}
        onUpdate={[Function]}
        required={false}
        ssn={Object {}}
        validator={[Function]}
        verified={false}
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ApplicantSSN {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "handleCompletion": [Function],
        "handleError": [Function],
        "props": Object {
          "dispatch": [Function],
          "onError": [Function],
          "onUpdate": [Function],
          "required": false,
          "ssn": Object {},
          "validator": [Function],
          "verified": false,
        },
        "refs": Object {},
        "section": "identification",
        "state": Object {
          "error": false,
          "uid": "undefined-MOCK-GUID",
          "verification": Object {},
        },
        "store": "Identification",
        "storeKey": "ApplicantSSN",
        "subsection": "ssn",
        "updateSSN": [Function],
        "updateVerification": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
        "verificationError": [Function],
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="section-content applicant-ssn"
          data-section="identification"
          data-subsection="ssn"
        >
          <h1
            className="section-header"
          >
            Social security number
          </h1>
          <Field
            adjustFor=""
            className=""
            comments={false}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help="identification.ssn.help"
            helpActive={false}
            optional={false}
            optionalText=""
            scrollIntoView={true}
            shrink={false}
            title="Provide your U.S. Social Security Number"
            titleSize="h4"
            validate={true}
          >
            <SSN
              className="applicant-ssn-initial"
              error={false}
              first=""
              focus={false}
              hideNotApplicable={false}
              last=""
              middle=""
              name="ssn"
              notApplicable={false}
              onError={[Function]}
              onUpdate={[Function]}
              required={false}
              tab={[Function]}
              valid={false}
              value=""
            />
          </Field>
          <Show
            when={false}
          >
            <Field
              adjustFor=""
              className=""
              comments={false}
              commentsActive={false}
              commentsAdd="comments.add"
              commentsName="Comments"
              commentsRemove="comments.remove"
              commentsRequired={false}
              commentsValue={Object {}}
              errorPrefix=""
              errors={Array []}
              filterErrors={[Function]}
              help=""
              helpActive={false}
              optional={false}
              optionalText=""
              scrollIntoView={true}
              shrink={false}
              title="Please confirm your social security number"
              titleSize="h4"
              validate={true}
            >
              <SSN
                className="applicant-ssn-verification"
                error={false}
                first=""
                focus={false}
                hideNotApplicable={true}
                last=""
                middle=""
                name="verification"
                notApplicable={false}
                onError={[Function]}
                onUpdate={[Function]}
                required={false}
                tab={[Function]}
                valid={false}
                value=""
              />
            </Field>
          </Show>
          <Show
            when={false}
          >
            <Field
              adjustFor=""
              className=""
              comments={false}
              commentsActive={false}
              commentsAdd="comments.add"
              commentsName="Comments"
              commentsRemove="comments.remove"
              commentsRequired={false}
              commentsValue={Object {}}
              errorPrefix=""
              errors={Array []}
              filterErrors={[Function]}
              help=""
              helpActive={false}
              optional={true}
              optionalText=""
              scrollIntoView={true}
              shrink={false}
              title="Thank you for confirming your social security number!"
              titleSize="h4"
              validate={true}
            />
          </Show>
        </div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="section-content applicant-ssn"
          data-section="identification"
          data-subsection="ssn"
        >
          <h1
            className="section-header"
          >
            Social security number
          </h1>
          <Field
            adjustFor=""
            className=""
            comments={false}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help="identification.ssn.help"
            helpActive={false}
            optional={false}
            optionalText=""
            scrollIntoView={true}
            shrink={false}
            title="Provide your U.S. Social Security Number"
            titleSize="h4"
            validate={true}
          >
            <SSN
              className="applicant-ssn-initial"
              error={false}
              first=""
              focus={false}
              hideNotApplicable={false}
              last=""
              middle=""
              name="ssn"
              notApplicable={false}
              onError={[Function]}
              onUpdate={[Function]}
              required={false}
              tab={[Function]}
              valid={false}
              value=""
            />
          </Field>
          <Show
            when={false}
          >
            <Field
              adjustFor=""
              className=""
              comments={false}
              commentsActive={false}
              commentsAdd="comments.add"
              commentsName="Comments"
              commentsRemove="comments.remove"
              commentsRequired={false}
              commentsValue={Object {}}
              errorPrefix=""
              errors={Array []}
              filterErrors={[Function]}
              help=""
              helpActive={false}
              optional={false}
              optionalText=""
              scrollIntoView={true}
              shrink={false}
              title="Please confirm your social security number"
              titleSize="h4"
              validate={true}
            >
              <SSN
                className="applicant-ssn-verification"
                error={false}
                first=""
                focus={false}
                hideNotApplicable={true}
                last=""
                middle=""
                name="verification"
                notApplicable={false}
                onError={[Function]}
                onUpdate={[Function]}
                required={false}
                tab={[Function]}
                valid={false}
                value=""
              />
            </Field>
          </Show>
          <Show
            when={false}
          >
            <Field
              adjustFor=""
              className=""
              comments={false}
              commentsActive={false}
              commentsAdd="comments.add"
              commentsName="Comments"
              commentsRemove="comments.remove"
              commentsRequired={false}
              commentsValue={Object {}}
              errorPrefix=""
              errors={Array []}
              filterErrors={[Function]}
              help=""
              helpActive={false}
              optional={true}
              optionalText=""
              scrollIntoView={true}
              shrink={false}
              title="Thank you for confirming your social security number!"
              titleSize="h4"
              validate={true}
            />
          </Show>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ApplicantSSN
    dispatch={[Function]}
    onError={[Function]}
    onUpdate={[Function]}
    required={false}
    ssn={Object {}}
    validator={[Function]}
    verified={false}
  />,
}
`;
